### PR TITLE
Separate properties and geom updates signals

### PR DESCRIPTION
**Pull request by LePetitTim**

### CODE REVIEW

File: terra_geocrud/serializers.py [hunk 1]
```diff
@@ -300,9 +300,17 @@ def update(self, instance, validated_data):
                                                             defaults={'route_description': route_description.get(
                                                                 'route_description',
                                                                 {})})
+        geom = validated_data.get("geom")
+        properties = validated_data.get("properties")
+        update_fields = []
+        if geom and geom != instance.geom:
+            update_fields = ['geom']
+        elif properties and properties != instance.properties:
+            update_fields = ['properties']
```
Comment (jrmi) [inline terra_geocrud/serializers.py:309]: Can't we have both: geom and properties ?

Comment (LePetitTim) [reply terra_geocrud/serializers.py:303]: We can't because of the front

File: terra_geocrud/signals.py [hunk 2]
```diff
@@ -6,23 +6,32 @@
...
 signals.post_save.disconnect(save_feature, sender=Feature)
 
 signals.post_save.disconnect(save_layer_relation, sender=Feature)
 
 
+def execute_async_save(update_fields, instance, kwargs):
+    if (update_fields is None or 'geom' in update_fields) and hasattr(
+            instance.layer, 'crud_view'):
+        execute_async_func(feature_update_relations_and_properties, (instance.pk, kwargs))
+    if "properties" in update_fields and hasattr(
```
Comment (gbip) [inline terra_geocrud/signals.py:22]: If `update_fields is None` means that this is a new object, shouldn't we test for it when updating properties ?

File: terra_geocrud/tasks.py [hunk 3]
```diff
@@ -36,37 +36,46 @@ def change_props(feature):
...
 
-def sync_relations_destination(feature, kwargs):
+def sync_properties_relations_destination(feature, update_relations=False):
     for relation_destination in feature.layer.relations_as_destination.all():
         for feature in relation_destination.origin.features.all():
-            feature.sync_relations(relation_destination.pk)
-            change_props(feature)
+            if update_relations:
+                feature.sync_relations(relation_destination.pk)
+            change_props(feature)  # TODO:  Compute link between relation and computed properties
```
Comment (gbip) [inline terra_geocrud/tasks.py:49]: Should this be in a for loop ?

Comment (LePetitTim) [reply terra_geocrud/tasks.py:49]: yes : 
Relation 1 with origin Trek and destination POIs
Relation 2 with origin Way and destination POIs
You have to loop on all origin features for each relations Relation 1 and Relation 2 to 

If you change 1 property of POI, and ways and treks have properties linked to your POI you have to change every single features of your treks and pois to synchronize their properties.

File: terra_geocrud/signals.py (hunk 2 above)
Comment (LePetitTim) [reply terra_geocrud/signals.py]: if update fields is None it means it's save from an other place than the view CrudFeatureViewSet. (Detail view)

File: terra_geocrud/tasks.py (hunk 3 above)
Comment (gbip) [reply terra_geocrud/tasks.py:49]: Ok, feature in the for loop is shadowing the argument, my bad.

File: terra_geocrud/tests/test_signals.py [hunk 4]
```diff
@@ -20,7 +21,7 @@ def side_effect_async(async_func, args=()):
         mocked.side_effect = side_effect_async
 
 
-@patch('terra_geocrud.tasks.feature_update_relations_destinations.delay')
+@patch('terra_geocrud.tasks.feature_update_relations_and_properties.delay')
```
Comment (gbip) [inline terra_geocrud/tests/test_signals.py:24]: Il faut patcher là où cette fonction est importée. Donc en fonction du code exercé par ton test il va falloir patcher  :
* `terra_geocrud.signals.execute_async_save`
* `terra_geocrud.tasks.feature_update_relations_origins`
* `terra_geocrud.tasks.layer_relations_set_destinations`
(ou les 3)

File: terra_geocrud/tests/test_signals.py [hunk 5]
```diff
@@ -373,10 +375,23 @@ def side_effect_async_origins(feature_id, kwargs):
...
+        def side_effect_async_destinations(feature_id, kwargs):
+            Feature.objects.get(pk=feature_id).delete()
+            task_result = feature_update_destination_properties(feature_id, kwargs)
+            assert not task_result
+
+        async_delay_destinations.side_effect = side_effect_async_destinations
+        property_mocked.return_value = True
+
+        self.add_side_effect_async(async_mocked)
+        self.feature_long.save(update_fields=['properties'])
```
Comment (gbip) [inline terra_geocrud/tests/test_signals.py:389]: Are you missing an assert or something in this test ?

Comment (LePetitTim) [reply terra_geocrud/tests/test_signals.py:389]: he is in the side effect, i check only that the task fail

File: terra_geocrud/tasks.py [hunk 6]
```diff
@@ -16,7 +16,7 @@ def compute_properties(instance, prop):
     instance.properties[prop.key] = value
     try:
         instance.clean()
-        instance.save(update_fields=['properties'])
+        Feature.objects.bulk_update([instance], ['properties'])
```
Comment (gbip) [inline terra_geocrud/tasks.py:19]: Un petit commentaire pour expliquer pourquoi tu fais un bulk update ? :smiley:

File: terra_geocrud/signals.py [hunk 7]
```diff
@@ -6,23 +6,31 @@
...
 signals.post_save.disconnect(save_feature, sender=Feature)
 
 signals.post_save.disconnect(save_layer_relation, sender=Feature)
 
 
+def execute_async_save(update_fields, instance, kwargs):
+    if update_fields is None or 'geom' in update_fields:
+        execute_async_func(feature_update_relations_and_properties, (instance.pk, kwargs))
+    elif "geom" not in update_fields and "properties" in update_fields:
+        execute_async_func(feature_update_destination_properties, (instance.pk, kwargs))
```
Comment (gbip) [inline terra_geocrud/signals.py:22]: Est-ce que je peux te suggérer de re-écrire tes conditions pour que ce soit plus clair les cas traités/non-traités par ce signaux ?

Par exemple :
```suggestion
    if update_fields :
        if "geom" not in update_fields and "properties" in update_fields :
           # partial save => properties updated
        elif "geom" in update_fields and "properties"  not in update_fields :
           # partial save => geom updated
        elif "geom" in update_fields and "properties" in update_fields :
           # partial save => properties & geom updated
        else :
           # ???
    else :
          # full update
```

File: terra_geocrud/tests/test_signals.py [hunk 8]
```diff
@@ -372,11 +374,85 @@ def side_effect_async_origins(feature_id, kwargs):
                           (0, 0)))
         )
         feature.delete()
+        async_delay_destinations.assert_called()
+        async_delay_origins.assert_called()
```
Comment (gbip) [inline terra_geocrud/tests/test_signals.py:378]:
```suggestion
        async_delay_origins.assert_called_once()
```
Is it relevant ?

Comment (LePetitTim) [reply terra_geocrud/tests/test_signals.py]: Will be call 3 and 2 times

File: terra_geocrud/signals.py (hunk 7 above)
Comment (LePetitTim) [reply terra_geocrud/signals.py]: I change for a if, else thanks to the bulk_update

File: terra_geocrud/tasks.py (hunk 6 above)
Comment (LePetitTim) [reply terra_geocrud/tasks.py:20]: Done

**Pull request merged**